src/web.security/middleware/index.ts

```diff
@@ -0,0 +1,80 @@
+import {NextFunction, Request, Response} from "express";
+import {COOKIE_NAME, OAUTH2_AUTH_URI, OAUTH2_CLIENT_ID,
+  OAUTH2_REDIRECT_URI} from "../../properties";
+import * as redisService from "../../services/redis.service";
+import * as keys from "../../session/keys";
+import logger from "../../logger";
+import {jweEncodeWithNonce, generateNonce} from "../jwt.encryption";
+
+const OAUTH_SCOPE_PREFIX = "https://api.companieshouse.gov.uk/company/";
+const COMPANY_NUMBER_KEY = "company_number";
+
+export default async (req: Request, res: Response, next: NextFunction) => {
+  // Get CompanyNumber from URI
+  const companyNumber = getCompanyFromPath(req.originalUrl);
+  if (companyNumber === "") {
+    return next(new Error("No Company Number"));
+  }
+
+  const cookieId = req.cookies[COOKIE_NAME];
+  if (cookieId) {
+    req.chSession = await redisService.loadSession(cookieId);
+    const signInInfo = req.chSession.getSignedInInfo();
+    if (isAuthorisedForCompany(signInInfo, companyNumber)) {
+        logger.info("User is authenticated for %s", companyNumber);
+        return next();
+    } else {
+      return res.redirect(await getAuthRedirectUri(req, companyNumber));
+    }
+  } else {
+    // TODO: Process no session
```

Comment:
Is there a JIRA task or story for this outstanding functionality? Would be good to create and reference it here.